esp-bootloader-esp-idf/src/ota.rs

```diff
@@ -0,0 +1,316 @@
+{
+    flash: &'a mut FlashRegion<'a, F>,
+}
+
+impl<'a, F> Ota<'a, F>
+where
+    F: embedded_storage::Storage,
+{
+    /// Create a [Ota] instance from the given [FlashRegion]
+    ///
+    /// # Errors
+    /// A [crate::partitions::Error::InvalidPartition] if the given flash region
+    /// doesn't represent a Data/Ota partition or the size is unexpected
+    pub fn new(flash: &'a mut FlashRegion<'a, F>) -> Result<Ota<'a, F>, crate::partitions::Error> {
+        if flash.capacity() != 0x2000
+            || flash.raw.partition_type()
+                != crate::partitions::PartitionType::Data(
+                    crate::partitions::DataPartitionSubType::Ota,
+                )
+        {
+            return Err(crate::partitions::Error::InvalidPartition {
+                expected_size: 0x2000,
+                expected_type: crate::partitions::PartitionType::Data(
+                    crate::partitions::DataPartitionSubType::Ota,
+                ),
+            });
+        }
+
+        Ok(Ota { flash })
+    }
+
+    /// Returns the currently active OTA-slot.
+    pub fn current_slot(&mut self) -> Result<Slot, crate::partitions::Error> {
+        let (seq0, seq1) = self.get_slot_seq()?;
+
+        let slot = if seq0 == 0xffffffff && seq1 == 0xffffffff {
+            Slot::None
+        } else if seq0 == 0xffffffff {
+            Slot::Slot1
+        } else if seq1 == 0xffffffff {
+            Slot::Slot0
+        } else if seq0 > seq1 {
+            Slot::Slot0
+        } else {
+            Slot::Slot1
+        };
+
+        Ok(slot)
+    }
+
+    fn get_slot_seq(&mut self) -> Result<(u32, u32), crate::partitions::Error> {
+        let mut buffer1 = OtaSelectEntry::default();
+        let mut buffer2 = OtaSelectEntry::default();
+        self.flash.read(SLOT0_DATA_OFFSET, buffer1.as_bytes_mut())?;
+        self.flash.read(SLOT1_DATA_OFFSET, buffer2.as_bytes_mut())?;
+        let seq0 = buffer1.ota_seq;
+        let seq1 = buffer2.ota_seq;
+        Ok((seq0, seq1))
+    }
+
+    /// Sets the currently active OTA-slot.
+    ///
+    /// Passing [Slot::None] will reset the OTA-data
+    pub fn set_current_slot(&mut self, slot: Slot) -> Result<(), crate::partitions::Error> {
+        if slot == Slot::None {
+            self.flash.write(SLOT0_DATA_OFFSET, &[0xffu8; 0x20])?;
+            self.flash.write(SLOT1_DATA_OFFSET, &[0xffu8; 0x20])?;
+            return Ok(());
+        }
+
+        let (seq0, seq1) = self.get_slot_seq()?;
+
+        let new_seq = {
+            if seq0 == 0xffffffff && seq1 == 0xffffffff {
+                1
+            } else if seq0 == 0xffffffff {
+                seq1 + 1
+            } else if seq1 == 0xffffffff {
+                seq0 + 1
+            } else {
+                u32::max(seq0, seq1) + 1
+            }
+        };
+
+        let crc = Crc::<u32>::new(&ALGO);
+        let mut digest = crc.digest();
+        digest.update(&new_seq.to_le_bytes());
+        let checksum = digest.finalize();
```

Comment:
I'm 60% sure there's a ROM function for this.

Reply:
https://github.com/esp-rs/esp-hal/blob/1db73628c24f6c9c7d3938b233ceecabee42fe31/esp-hal/src/rom/crc.rs#L132

Reply:
I was originally thinking of having host tests but for multiple reasons I didn't - so we could use the ROM function actually